sdk/azidentity/device_code_credential.go

```diff
@@ -0,0 +1,100 @@
+// options: allow to configure the management of the requests sent to the Azure Active Directory service.
+func NewDeviceCodeCredential(tenantID string, clientID string, callback func(string), options *TokenCredentialOptions) *DeviceCodeCredential {
+	return &DeviceCodeCredential{tenantID: tenantID, clientID: clientID, callback: callback, client: newAADIdentityClient(options)}
+}
+
+// GetToken obtains a token from the Azure Active Directory service, following the device code authentication
+// flow. This function first requests a device code and requests that the user login to continue authenticating.
+// This function will keep polling the service for a token meanwhile the user logs.
+// - scopes: The list of scopes for which the token will have access.
+// - ctx: controlling the request lifetime.
+// Returns an AccessToken which can be used to authenticate service client calls.
+func (c *DeviceCodeCredential) GetToken(ctx context.Context, opts azcore.TokenRequestOptions) (*azcore.AccessToken, error) {
+	if len(c.refreshToken) != 0 {
+		tk, err := c.client.refreshAccessToken(ctx, c.tenantID, c.clientID, "", c.refreshToken, opts.Scopes)
+		if tk == nil { // this would only happen if there was an unmarshal error or a failure to parse ExpiresOn
+			return nil, err
+		}
+		// assign new refresh token to the credential for future use
+		c.refreshToken = tk.refreshToken
+		// passing the access token and/or error back up
+		return tk.token, err
+	}
+	// if there is no refreshToken, then begin the Device Code flow from the beginning
+	// make initial request to the device code endpoint for a device code and instructions for authentication
+	dc, err := c.client.requestNewDeviceCode(ctx, c.tenantID, c.clientID, opts.Scopes)
+	if err != nil {
+		return nil, err
+	}
+	// send authentication flow instructions back to the user to log in and authorize the device
+	c.callback(dc.Message)
+	// poll the token endpoint until a valid access token is received or until authentication fails
+	for {
+		tk, err := c.client.authenticateDeviceCode(ctx, c.tenantID, c.clientID, dc.DeviceCode, opts.Scopes)
+		// if there is no error, save the refresh token and return the token credential
+		if err == nil {
+			c.refreshToken = tk.refreshToken
+			return tk.token, err
+		}
+		// if there is an error, check for an AADAuthenticationFailedError in order to check the status for token retrieval
+		var authRespErr *AADAuthenticationFailedError
+		// if the error is not an AADAuthenticationFailedError, then fail here since something unexpected occurred
+		if !errors.As(err, &authRespErr) {
+			return nil, err
+		}
+		switch authRespErr.Message {
+		// wait for the interval specified from the initial device code endpoint and then poll for the token again
+		case "authorization_pending":
+			time.Sleep(time.Duration(dc.Interval) * time.Second)
+		default:
+			// any other error should be returned
+			return nil, err
```

Comment:
Shouldn't this return an `AuthenticationFailedError` to be consistent with the other credential types?

Reply:
Yes

Reply:
Update: this is actually how it should be because the wrapping is carried out within authenticateDeviceCode. The errors that could be returned are json unmarshal errors, pipeline errors OR AuthenticationFailedErrors. So, when this is in fact an AuthenticationFailedError the wrapping will be there. See line: 
https://github.com/Azure/azure-sdk-for-go/pull/6560/files#diff-a7c4c602cbc8f32f6e9e22430ee205e1R316